### PR TITLE
Receiver: return after timeout even if more data is pending

### PR DESCRIPTION
**Pull request by eile**

### CODE REVIEW

File: doc/Changelog.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 # git master
 
+* [218](https://github.com/HBPVIS/ZeroEQ/pull/218):
+  Fix infinite loop in slow receivers
```
Comment (tribal-tec) [inline doc/Changelog.md]: more like "with very fast sender(s)"

File: zeroeq/receiver.cpp [hunk 2]
```diff
@@ -106,13 +113,18 @@ class Receiver
                 intervals.push_back(sockets.size() - before);
             }
 
-            switch (zmq_poll(sockets.data(), int(sockets.size()), timeout))
+            const uint32_t remaining =
+                duration_cast<milliseconds>(high_resolution_clock::now() -
+                                            startTime)
+                    .count();
```
Comment (tribal-tec) [inline zeroeq/receiver.cpp]: nice formatting :)

**Comment by rdumusc** [comment]
@eile @tribal-tec this commit breaks the http server in Tide, which is based on QSocketNotifier. I don't understand what you were trying to achieve here. Doing a non-blocking receive() using timeout=0 is not allowed anymore? Please advise.

**Comment by eile** [comment]
Seems there is no test for this. :P

This PR fixed that if the publisher is faster than the subscriber, a subscriber.receive(timeout) would never return, even after the timeout expired.

This changes the behaviour of receive(0) which now will do one receive, not all (which is in line with the doc/spec). 

Not sure why/how this breaks the QSocketNotifier.

**Comment by rdumusc** [comment]
OK, part of the problem is that the timeout is now used for two purposes:
- a poll timeout (waiting x milliseconds until SOME data arrives)
- a processing timeout (abort after processing TOO MUCH data)

I think the reason it breaks in Tide is because the zmq socket notifies only once data first arrives (edge triggered), not *as long as* there is data to read (level triggered). So now it only processes one message, exceeds the limit of 0 millisecond and stops instead of reading all messages.

**Comment by tribal-tec** [comment]
Wouldn't ```while(receive(0));``` solve it then in Tide to process all pending?

**Comment by rdumusc** [comment]
seeing that receive() returns false when no messages were processed probably yes, I will test that next week

**Pull request merged**